### ISSUE TITLE
[Installation]: Is there a way to put in a yt-dlp.conf file to enable a socks5 proxy

### ISSUE THREAD

**Issue by mwexcell** [tubearchivist/tubearchivist#818]
### Installation instructions

- [X] I have read the [how to open an issue](https://github.com/tubearchivist/tubearchivist/blob/master/CONTRIBUTING.md#how-to-open-an-issue) guide, particularly the [installation help](https://github.com/tubearchivist/tubearchivist/blob/master/CONTRIBUTING.md#installation-help) section.

### Operating System

Linux

### Your installation or setup question

First off, incredible system - much more usable and stable than competitive systems.

In other systems that I have used that have yt-dlp under the hood, I can either configure parameters that are included in the yt-dlp configuration when the library is called, or command line arguments that it will pass if it uses yt-dlp as a system process, or I can drop in an /etc/yt-dlp.conf or other conf file that gets automatically picked up (automatically by yt-dlp - not the app). Is there any way to do something similar with Tube Archivist so that I can enable my socks5 proxy in yt-dlp?

### Relevant log output

```Shell
N/A
```

**Comment by 7677865466** [comment]
You can pass these environment variables through docker and podman. Don't forget to exclude communication between tubearchivist and both the archivist-es and archivist-redis containers from using the proxy. I am using this myself in combination with a Cloudflare-Warp socks5 proxy.

https://docs.docker.com/engine/daemon/proxy/

**Comment by kivle** [comment]
Hi,

I've been looking into this same issue today. I was looking for ways I could make the downloading and scraping run from behind a vpn and/or a proxy.

Sorry for the wall of text. Feel free to skip to the tldr;

I first tried putting the TubeArchivist container through a vpn container by using `network_mode: "container:my_vpn_container_name"`. 

```yaml
services:
  tubearchivist:
    container_name: tubearchivist
    restart: unless-stopped
    image: bbilly1/tubearchivist
    network_mode: container:gluetun
   ...
```

While this puts TA behind the vpn, it also completely isolates it from Redis and ElasticSearch containers, so it's not really a viable solution. I've been looking for possible workarounds for this, but I've not been able to figure anything out.

But gluetun (the vpn container I use) also provides an HTTP Proxy, so then I started looking into the possibility of getting yt_dlp to use that through the `--proxy` parameter. From what I can gather the CLI for yt_dlp can either take command line parameters, or [parameter overrides in config files](https://github.com/yt-dlp/yt-dlp?tab=readme-ov-file#configuration). It can _not_ take any parameter overrides from environment variables. So as a test I created a tiny `yt-dlp.conf` file and mounted it under `/etc/yt-dlp.conf` and added the gluetun network under `networks` for the TA container.

```
--proxy http://gluetun:8888
```

But this turned out to be another dead end.

tldr; As a library, yt_dlp will _only_ use the parameters you pass directly to it in the dictionary. No command line parameters or config files will be parsed. But I think I have found a very simple workaround. I was not able to test it since I do not know how to set up the dev environment properly to test this locally. I have attached the simple one-liner diff below:

```diff
diff --git a/tubearchivist/home/src/download/yt_dlp_base.py b/tubearchivist/home/src/download/yt_dlp_base.py
index d95a297..6917e07 100644
--- a/tubearchivist/home/src/download/yt_dlp_base.py
+++ b/tubearchivist/home/src/download/yt_dlp_base.py
@@ -34,6 +34,7 @@ class YtWrap:
     def build_obs(self):
         """build yt-dlp obs"""
         self.obs = self.OBS_BASE.copy()
+        self.obs.update(yt_dlp.parse_options())
         self.obs.update(self.obs_request)
         if self.config:
             self.add_cookie()
```

I believe this small change should make yt_dlp search for and parse the config files.

**Comment by bbilly1** [comment]
Proxy/VPN/Tunneling needs to be handled outside of this project, see the FAQ for reference [here](https://docs.tubearchivist.com/faq/#how-do-i-tunnel-all-traffic-from-this-container).

Having a yt-dlp config file as suggested is very likely going to break things, as you can easily reach an unexpected config that TA doesn't know how to handle. All config options exposed to the user are accessible on the settings page.

When setting up gluetun for example, you'll probably need to define an internal network for the container dependencies. But that is outside of the scope of this, you'll have to refer to the gluetune docs respectively.

**Comment by PawsFunctions** [comment]
can you please add extra option in advance settings for proxy IP:port

thank you

**Comment by culdev** [comment]
According to yt-dlp config you should be able to configure proxy via environment variable (YTDLP_PROXY) but I couldn't get it to work at all

I then tried the HTTP_PROXY environment variable (Python urllib), it worked but then it started proxying calls to elasticsearch too which I didn't want

So now I'm trying out same approach as noted above, I edit yt_dlp_base.py with an entrypoint script, it's been working well so far, but still feels dirty, we'll see how long I will be using it...

Anyways, I did this, in docker-compose.yml I added my script and entrypoint:
```yml
services:
  tubearchivist:
    container_name: tubearchivist
    restart: always
    image: bbilly1/tubearchivist
    entrypoint: /applyproxy.sh
    ports:
      - 8001:8000
    volumes:
      - tubearchivist-media:/youtube
      - ./data/tubearchivist/cache:/cache
      - ./applyproxy.sh:/applyproxy.sh:ro
```

and applyproxy.sh looks like this (don't forget to set it as executable `chmod +x applyproxy.sh`):

```sh
#!/bin/bash

echo "Applying special proxy crap"

# Choose one of the following sed:
sed -i 's; "retries": 10,;"proxy": "http://172.27.0.5:8080","retries": 10,;g' /app/download/src/yt_dlp_base.py

# Had trouble with "The page needs to be reloaded." error
# Found a solution https://github.com/yt-dlp/yt-dlp/issues/17405#issuecomment-5248997656
# So I added extractor args here
#sed -i 's; "retries": 10,;"proxy": "http://172.27.0.5:8080","extractor_args": {"youtube": {"player_client": ["web_safari","web_embedded","-tv_downgraded"]}},"retries": 10,;g' /app/download/src/yt_dlp_base.py

echo "Starting tubearchivist..."
/app/run.sh
```

Sed will look for ` "retries": 10,` in `yt_dlp_base.py` and add proxy in front of it, because of the space in the beginning it will only happen once.
Change `http://172.27.0.5:8080` to whatever proxy you want to use that ytp-dl supports (http/https/socks5)
Note don't use `;` in the url, it will break the sed pattern

**Comment by seproDev** [comment]
> you should be able to configure proxy via environment variable (YTDLP_PROXY)

Where did you find this information? No such envirment variable exists in yt-dlp.

**Comment by culdev** [comment]
> > you should be able to configure proxy via environment variable (YTDLP_PROXY)
> 
> Where did you find this information? No such envirment variable exists in yt-dlp.

It was mostly a hope that all CLI options had a environment variable name and a guess from https://github.com/yt-dlp/yt-dlp?tab=readme-ov-file#notes-about-environment-variables
and the following mentions of "environment variable"
> Set the environment variable YTDLP_NO_PLUGINS to something nonempty to disable loading plugins entirely.
> Set the environment variable YTDLP_NO_LAZY_EXTRACTORS to something nonempty to forcefully disable lazy extractor loading.

**Comment by PawsFunctions** [comment]
when yt-dlp is used as library, we can't use any config file or environment variable if its not told to do it.
` self.obs.update(yt_dlp.parse_options().ydl_opts)`

@culdev  thanks for better solution.

**Comment by jelev** [comment]
> According to yt-dlp config you should be able to configure proxy via environment variable (YTDLP_PROXY) but I couldn't get it to work at all
> 
> I then tried the HTTP_PROXY environment variable (Python urllib), it worked but then it started proxying calls to elasticsearch too which I didn't want
> 
> So now I'm trying out same approach as noted above, I edit yt_dlp_base.py with an entrypoint script, it's been working well so far, but still feels dirty, we'll see how long I will be using it...
> 
> Anyways, I did this, in docker-compose.yml I added my script and entrypoint:
> 
> services:
>   tubearchivist:
>     container_name: tubearchivist
>     restart: always
>     image: bbilly1/tubearchivist
>     entrypoint: /applyproxy.sh
>     ports:
>       - 8001:8000
>     volumes:
>       - tubearchivist-media:/youtube
>       - ./data/tubearchivist/cache:/cache
>       - ./applyproxy.sh:/applyproxy.sh:ro
> 
> and applyproxy.sh looks like this (don't forget to set it as executable `chmod +x applyproxy.sh`):
> 
> #!/bin/bash
> 
> echo "Applying special proxy crap"
> sed -i 's; "retries": 10,;"proxy": "http://172.27.0.5:8080","retries": 10,;g' /app/download/src/yt_dlp_base.py
> 
> echo "Starting tubearchivist..."
> /app/run.sh
> 
> Sed will look for ` "retries": 10,` in `yt_dlp_base.py` and add proxy in front of it, because of the space in the beginning it will only happen once. Change `http://172.27.0.5:8080` to whatever proxy you want to use that ytp-dl supports (http/https/socks5) Note don't use `;` in the url, it will break the sed pattern

version 5.1

`find /app/ -iname 'yt_dlp_base.py' -exec sed -i 's; "retries": 10,;"proxy": "http://172.27.0.5:8080","retries": 10,;g' {} \;`
